### DISCUSSION TITLE
How to test rich to use __rich__ when logging objects?

### DISCUSSION THREAD

**Discussion by ssbarnea** [Textualize/rich#1154] [Q&A]
Based on the documentation I had the expectation that once I add `__rich__` method to a class, printing those objects with rich would work, including for logging.  Still, when trying to log an object that implemented __rich__ protocol, I still see it using the plain __str__ method.

```
logging.debug("[red]foo[/] %s detected !", myobject, extra={"markup": True})
# if I write myobject.__rich__() instead, it would work as desired
```

With example above I get foo as red but myobject is rendered as string instead of using the markup:

```
    def __rich__(self):
        return f"[green]{self.name}[/]"
```

What did I miss? -- If I use `__str__` instead it will work but that would also have the undesired effect of always returning the markup representation which is not desired.

**Comment by willmcgugan** [comment]
The positional args are used to build a string before it gets to Rich. See the [logging](https://docs.python.org/3/library/logging.html#logging.Logger.debug) docs for details.

**Reply by ssbarnea** [reply]
I was hoping to see an example on https://rich.readthedocs.io/en/latest/logging.html page about how to nicely log objects that implement __rich__, if possible.

Does this mean that the only way to achieve this is to subclass the logging.Logger() ?

**Reply by willmcgugan** [reply]
You're out of luck. The logging library was designed to work with strings. You _could_ pass a Rich object via the `extra` argument and print that, but obviously that could only work via a customized RichHandler.

Having different output per handler is somewhat at odds with the goal of logging which can be configured to write logs to a variety of destinations, most of which won't be able to deal with Rich's formatting.
